esp/esdllib/esdl_script.cpp

```diff
@@ -0,0 +1,914 @@
+        return new CEsdlTransformOperationChoose(element, prefix);
+    if (strieq(op, "for-each"))
+        return new CEsdlTransformOperationForEach(element, prefix);
+    if (strieq(op, "if"))
+        return new CEsdlTransformOperationConditional(element, prefix);
+    if (streq(op, "set-value") || streq(op, "SetValue"))
+        return new CEsdlTransformOperationSetValue(element, prefix);
+    if (streq(op, "append-to-value") || streq(op, "AppendValue"))
+        return new CEsdlTransformOperationAppendValue(element, prefix);
+    if (streq(op, "add-value"))
+        return new CEsdlTransformOperationAddValue(element, prefix);
+    if (streq(op, "fail"))
+        return new CEsdlTransformOperationFail(element, prefix);
+    if (streq(op, "assert"))
+        return new CEsdlTransformOperationAssert(element, prefix);
+    return nullptr;
+}
+
+static IPropertyTree *getTargetPTree(IPropertyTree *tree, IXpathContext *xpathContext, const char *target)
+{
+    StringBuffer xpath(target);
+    if (xpath.length())
+    {
+        //we can use real xpath processing in the future, for now simple substitution is fine
+        StringBuffer variable;
+        xpath.replaceString("{$query}", xpathContext->getVariable("query", variable));
+        xpath.replaceString("{$method}", xpathContext->getVariable("method", variable.clear()));
+        xpath.replaceString("{$service}", xpathContext->getVariable("service", variable.clear()));
+        xpath.replaceString("{$request}", xpathContext->getVariable("request", variable.clear()));
+
+        IPropertyTree *child = tree->queryPropTree(xpath.str());  //get pointer to the write-able area
+        if (!child)
+            throw MakeStringException(-1, "EsdlCustomTransform error getting target xpath %s", xpath.str());
+        return child;
+    }
+    return tree;
+}
+static IPropertyTree *getOperationTargetPTree(MapStringToMyClass<IPropertyTree> &treeMap, IPropertyTree *currentTree, IEsdlTransformOperation &operation, IPropertyTree *tree, IXpathContext *xpathContext, const char *target)
+{
+    const char *mergedTarget = operation.queryMergedTarget();
+    if (isEmptyString(mergedTarget) || streq(mergedTarget, target))
+        return currentTree;
+    IPropertyTree *opTree = treeMap.getValue(mergedTarget);
+    if (opTree)
+        return opTree;
+    opTree = getTargetPTree(tree, xpathContext, mergedTarget);
+    if (opTree)
+        treeMap.setValue(mergedTarget, LINK(opTree));
+    return opTree;
+}
+
+class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
+{
+private:
+    IArrayOf<IEsdlTransformOperation> m_variables; //keep separate and only at top level for now
+    IArrayOf<IEsdlTransformOperation> m_operations;
+    StringAttr m_name;
+    StringAttr m_target;
+    StringBuffer m_prefix;
+
+public:
+    CEsdlCustomTransform(){}
+    CEsdlCustomTransform(IPropertyTree &tree, const char *ns_prefix) : m_prefix(ns_prefix)
+    {
+        if (m_prefix.length())
```

Comment:
A non empty prefix could be passed in in the future if the xmlns were inherited from somewhere above.  Like a collection of scripts or even higher.  Generally I wouldn't use it and would have the xmlns at the top of the script.  I picture us storing scripts in individual files and including them in the manifest so should usually be self contained.  I put this here as a convenience in case it's needed.  I could remove the ability now, or If it doesn't get used when refactoring the binding level structure then I will remove.